apps/src/aichat/aichatApi.ts

```diff
@@ -101,3 +125,133 @@ export async function postAichatCheckSafety(
 
   return await response.json();
 }
+
+interface StartChatCompletionResponse {
+  requestId: number;
+  pollingIntervalMs: number;
+  backoffRate: number;
+}
+
+export interface GetChatRequestResponse {
+  executionStatus: ValueOf<typeof AiRequestExecutionStatus>;
+  response: string;
+}
+
+/**
+ * Perform chat completion by initiating an asynchronous request and polling for the response.
+ */
+async function chatCompletionAsyncPolling(
+  newMessage: ChatMessage,
+  storedMessages: ChatMessage[],
+  aichatModelCustomizations: AichatModelCustomizations,
+  aichatContext: AichatContext,
+  maxPollingTimeMs = MAX_POLLING_TIME_MS
+): Promise<ChatCompletionApiResponse> {
+  const payload = {
+    newMessage,
+    storedMessages,
+    aichatModelCustomizations,
+    aichatContext,
+  };
+
+  const response = await HttpClient.post(
+    `${ROOT_URL}/start_chat_completion`,
+    JSON.stringify(payload),
+    true,
+    {
+      'Content-Type': 'application/json; charset=UTF-8',
+    }
+  );
+
+  const {
+    requestId,
+    pollingIntervalMs,
+    backoffRate: serverBackoffRate,
+  } = (await response.json()) as StartChatCompletionResponse;
+
+  const startTime = Date.now();
+  const backoffRate = serverBackoffRate || 1; // In case the server doesn't return a backoff rate
+
+  let executionStatus: ValueOf<typeof AiRequestExecutionStatus> =
+    AiRequestExecutionStatus.NOT_STARTED;
+  let currentInterval = Math.max(pollingIntervalMs, MIN_POLLING_INTERVAL_MS);
+  let modelResponse: string = '';
+
+  while (
+    executionStatus < AiRequestExecutionStatus.SUCCESS &&
+    Date.now() - startTime < maxPollingTimeMs
+  ) {
+    await new Promise(resolve => setTimeout(resolve, currentInterval));
+    const chatResponse = await HttpClient.fetchJson<GetChatRequestResponse>(
+      `${ROOT_URL}/chat_request/${requestId}`
+    );
+    executionStatus = chatResponse.value.executionStatus;
+    modelResponse = chatResponse.value.response;
+    currentInterval *= backoffRate;
+  }
+
+  if (executionStatus < AiRequestExecutionStatus.SUCCESS) {
+    // Timed out
+    throw new Error('Chat completion request timed out');
+  }
+
+  return {
+    messages: getUpdatedMessages(newMessage, modelResponse, executionStatus),
+  };
+}
+
+/**
+ * Get the updated user and assistant message based on the status of the chat completion request.
+ */
+function getUpdatedMessages(
```

Comment:
This is the same logic that happens on our backend currently, just moving it here since the `AichatRequest` stores only the bot response, not the newly updated user and bot messages

(Context: this was an intentional decision since I felt it would be clearer for the `AichatRequest` model to only store the actual response and status code, rather than need to know how to compute the updated list of messages, which felt like a front-end specific structure. However, this API still returns a list of updated messages to the caller, so the rest of our front end continues to operate the same way).